---
title: "Mendelian Randomization Report"
subtitle: "Mendelian Randomization analysis [Smoking and Graves' disease (GD)]"
author: 
- name: "Mahdi Akbarzadeh - Email: akbarzadeh.ms@gmail.com - mob: (+98)9126490780"
- name: "Maryam Pourganji"
date: "`r Sys.Date()`"
fontsize: 30pt  
output: 
  rmarkdown::html_document:
    theme: journal


      
---


```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# Mendelian Randomization Report [Smoking and Graves' disease (GD)] {.tabset}

 Exposure 1: Smoking Initiation on Graves’ disease (GD)
 
 Exposure 2: Age of Initiation Smoking on Graves’ disease (GD)
 
 Exposure 3: Cigarettes per day on Graves’ disease (GD)
 
 Exposure 4: Life time smoking on Graves’ disease (GD)


## Exposure 1



### Introduction

* **Title:** Investigating the causality between **Smoking Initiation on GD**

    + Exposure: Smoking initiation, [GSCAN consortium](https://genome.psych.umn.edu/index.php/GSCAN) | Reference paper:  [Liu Paper 2019](https://www.nature.com/articles/s41588-018-0307-5) | GWAS ID:[GCST007458](https://www.ebi.ac.uk/gwas/studies/GCST007458) 
+ **Sample size**:341,427
   + Outcome: Graves disease (GD) Disease, [NBC Human Database](https://humandbs.biosciencedbc.jp/en/hum0197-v3-220) | Reference paper: [Saori Sakaue Paper 2022](https://pubmed.ncbi.nlm.nih.gov/34594039/) | GWAS ID: [GCST90018847](https://www.ebi.ac.uk/gwas/studies/GCST90018847)
+ **Sample size**: 458,620 , **Number of cases**: 1,678 , **Number of controls**: 456,942
  + Ancestry: **European** 
    
```{r setwd and packages, echo=FALSE, message=FALSE, warning=FALSE}
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
rm(list=ls())
library(TwoSampleMR)
library(MRPracticals)
library(phenoscanner)
library(MRPRESSO)
library(data.table)
library(writexl)
library(MendelianRandomization)
```

### Data Preparation
1- **Number of total SNPs in exposure:** 25,008 SNPs (<$5 \times 10^-8$)

2- **Number of SNPs exposure with p-value <** $5 \times 10^-8$: 7,846 SNPs

3- **Number of SNPs exposure after clumping **: 93 SNPs

4- **Number of total SNPs in outcome**: 23,868 SNPs

5- **Number of common variants between exposure and outcome**: 90 SNPs ("rs10114490" "rs72896886" "rs56820925  have been eliminated)

6- **Number of SNPs after replacing proxies**: 3 SNPs from [NIH LDproxy database](https://ldlink.nih.gov/?tab=ldproxy) according to EUR ancestry have been selected: We replace "rs10114490" , "rs72896886" and "rs56820925  by rs7871108&rs72896891&rs57703976 with R2 1 & 0.91 & 1, respectively).So, 93 SNPs remained.   

7- **Number of SNPs after harmonization (action=2) ** = 84 SNPs (Removing the following SNPs for incompatible alleles:
rs13246563, rs72896886 and Removing the following SNPs for being palindromic with intermediate allele frequencies:
rs10956809, rs1160685, rs13246563, rs2186122, rs6508144, rs7585579, rs7921378, rs9540729)

8- **Number of SNPs after removing HLA region with exploring in [HLA Genes, Nomenclature](https://hla.alleles.org/genes/index.html)** = 84 SNP 

9- **Number of SNPs after removing those that have MAF < 0.01** = 84 SNPs 

10- Checking pleiotropy by [PhenoScanner](http://www.phenoscanner.medschl.cam.ac.uk/):

How many SNPs have been eliminated after checking the PhenoScanner website: 84 SNPs



### Checking weakness of the instruments:

```{r Checking weekness, message=FALSE, warning=FALSE}
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_smokingInitiation_graves_AFTER_LD.txt")
data$F<-(data$beta.exposure/data$se.exposure)^2
summary(data$F)
```
How many SNPs have been eliminated with checking the weakness: 0 SNP


### RUN an initial MR:

```{r RUN an initial MR, message=FALSE, warning=FALSE}
res<-mr(data)
res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)	
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Testing Outlier with PRESSO test

```{r PRESSO, echo=FALSE, message=FALSE, warning=FALSE}

#Radial test
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_smokingInitiation_graves_AFTER_LD.txt")
data <- as.data.frame(data)

mr_presso(BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", SdOutcome = "se.outcome", SdExposure = "se.exposure",
          OUTLIERtest = TRUE, DISTORTIONtest = TRUE, data = data, NbDistribution = 1000,  SignifThreshold = 0.05)

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```


### Studentized residuals:

```{r Studentized residuals,  message=FALSE, warning=FALSE}
reg_1<-lm(data$beta.outcome~data$beta.exposure-1)
data$st_1<-rstandard(reg_1)

#Histogram plot
hist(data$st_1)
```

### Radial test

```{r Radial, echo=FALSE, message=FALSE, warning=FALSE}
#Radial test: 

radial_data<-format_radial(data$beta.exposure, data$beta.outcome,
                           data$se.exposure, data$se.outcome,
                           data$SNP)
ivw_model<-ivw_radial(radial_data, 0.05/nrow(radial_data),3,0.0001)
ivw_model$outliers
```


### Cook's distance

In statistics, Cook's distance or Cook's D is a commonly used estimate of the influence of a data point when performing a least-squares regression analysis.[1] In a practical ordinary least squares analysis, Cook's distance can be used in several ways:

1- To indicate influential data points that are particularly worth checking for validity.

2- To indicate regions of the design space where it would be good to be able to obtain more data points. 

**It is named after the American statistician R. Dennis Cook, who introduced the concept in 1977.**

[Refernce](https://en.wikipedia.org/wiki/Cook%27s_distance)

```{r Cook distance, message=FALSE, warning=FALSE}
par(mfrow = c(2, 2))
model <- lm(data$beta.outcome~data$beta.exposure-1)
plot(model)
par(mfrow = c(1, 1))

cooksD <- cooks.distance(model)
influential <- cooksD[(cooksD > (3 * mean(cooksD, na.rm = TRUE)))]
influential
```




### Run After deleting new outlier: Final Results:

```{r Final Results, message=FALSE, warning=FALSE}

data <- data[(data$SNP!="rs9423279" & data$SNP!="rs12027999" 
                 & data$SNP!="rs7631735" & data$SNP!="rs12042107" & data$SNP!="rs12186738" & 
                   data$SNP!="rs6433897" & data$SNP!="rs72789632" & data$SNP!="rs6728726" & data$SNP!="rs4674993" &
                   data$SNP!="rs7224742" & data$SNP!="rs301805" & data$SNP!="rs4236259" & data$SNP!="rs4571506" 
                 & data$SNP!="rs1445649" & data$SNP!="rs7197072" & data$SNP!="rs3001723" & data$SNP!="rs4759228" 
                 & data$SNP!="rs7929518" & data$SNP!="rs10233018"),]

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)


#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#scatter plot
p1 <- mr_scatter_plot(res, data); p1[[1]]

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Sensitivity analyses with MendelianRandomization Package

```{r MendelianRandomization, echo=FALSE, message=FALSE, warning=FALSE}

MRInputObject <- mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome)
IVWObject <- mr_ivw(MRInputObject,
                    model = "default",
                    robust = FALSE,
                    penalized = FALSE,
                    correl = FALSE,
                    weights = "simple",
                    psi = 0,
                    distribution = "normal",
                    alpha = 0.05)
IVWObject

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")
MRAllObject_all
mr_plot(MRAllObject_all)

mr_forest(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
          snp_estimates=FALSE,
          methods = c("ivw", "median", "wmedian", "egger", "maxlik", "mbe"))
###########
data$samplesize.exposure <- "341427"
data$samplesize.outcome <- "458620"
data$samplesize.exposure <- as.numeric(data$samplesize.exposure)
data$samplesize.outcome <- as.numeric(data$samplesize.outcome)

out <- directionality_test(data)
knitr::kable(out)

mr_steiger(
  p_exp = data$pval.exposure, 
  p_out = data$pval.outcome, 
  n_exp = data$samplesize.exposure, 
  n_out = data$samplesize.outcome, 
  r_xxo = 1, 
  r_yyo = 1,
  r_exp=0,
  r_out = 0.5
)
```

### Working with  MRraps

```{r MRraps, echo=FALSE, message=FALSE, warning=FALSE}

library(mr.raps)
mr.raps.simple(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome, diagnosis = FALSE)
mr.raps.all(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome)
mr_lasso(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_cML(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), MA = TRUE, DP = FALSE,, n = 458620)
mr_divw(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_mbe(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), weighting="weighted", stderror="delta",iterations=100)
mr_funnel(MRInputObject, CI = TRUE)
mr_plot(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
        error = TRUE, orientate = FALSE, line = "ivw", interactive = FALSE)
```



## Exposure 2

### Introduction

* **Title:** Investigating the causality between **Age Of Initiation Smoking on GD**

    + Exposure:Age of Initiation Smoking, [GSCAN consortium](https://genome.psych.umn.edu/index.php/GSCAN) | Reference paper:  [Liu Paper 2019](https://www.nature.com/articles/s41588-018-0307-5) | GWAS ID:[GCST007458](https://www.ebi.ac.uk/gwas/studies/GCST007458) 
    + **Sample size**: 341,427 
    + Outcome: Graves disease (GD) Disease, [NBC Human Database](https://humandbs.biosciencedbc.jp/en/hum0197-v3-220) | Reference paper: [Saori Sakaue Paper 2022](https://pubmed.ncbi.nlm.nih.gov/34594039/) | GWAS ID: [GCST90018847](https://www.ebi.ac.uk/gwas/studies/GCST90018847)
    + **Sample size**: 458,620 , **Number of cases**: 1,678 , **Number of controls**: 456,942
    + Ancestry: **European** 
    

### Data Preparation
1- **Number of total SNPs in exposure:** 2,076 SNPs

2- **Number of SNPs exposure with p-value <** $10^-8$ = 447 SNPs

3- **Number of SNPs exposure after clumping ** = 7 SNPs

4- **Number of total SNPs in outcome**: 1,968 SNPs

5- **Number of common variants between exposure and outcome**:7

6- **Number of SNPs after replacing proxies**: - 

7- **Number of SNPs after harmonization (action=2) ** = 7 SNPs 

8- **Number of SNPs after removing HLA region with exploring in [HLA Genes, Nomenclature](https://hla.alleles.org/genes/index.html)** = 7 SNP 

9- **Number of SNPs after removing those that have MAF < 0.01** = 7 SNPs 

10- Checking pleiotropy by [PhenoScanner](http://www.phenoscanner.medschl.cam.ac.uk/):

How many SNPs have been eliminated after checking the PhenoScanner website: 7 SNPs



### Checking weakness of the instruments:

```{r Checking weekness1, message=FALSE, warning=FALSE}
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_ageOfInitiation_graves_AFTER_LD.txt")
data$F<-(data$beta.exposure/data$se.exposure)^2
summary(data$F)
```
How many SNPs have been eliminated with checking the weakness: 0 SNP


### RUN an initial MR:

```{r RUN an initial MR1, message=FALSE, warning=FALSE}
res<-mr(data)
res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)	
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Testing Outlier with PRESSO test

```{r PRESSO1, echo=FALSE, message=FALSE, warning=FALSE}

#Radial test
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_ageOfInitiation_graves_AFTER_LD.txt")
data <- as.data.frame(data)

mr_presso(BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", SdOutcome = "se.outcome", SdExposure = "se.exposure",
          OUTLIERtest = TRUE, DISTORTIONtest = TRUE, data = data, NbDistribution = 1000,  SignifThreshold = 0.05)

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```


### Studentized residuals:

```{r Studentized residuals1, message=FALSE, warning=FALSE}
reg_1<-lm(data$beta.outcome~data$beta.exposure-1)
data$st_1<-rstandard(reg_1)

#Histogram plot
hist(data$st_1)
```

### Radial test

```{r Radial1, echo=FALSE, message=FALSE, warning=FALSE}
#Radial test: 

radial_data<-format_radial(data$beta.exposure, data$beta.outcome,
                           data$se.exposure, data$se.outcome,
                           data$SNP)
ivw_model<-ivw_radial(radial_data, 0.05/nrow(radial_data),3,0.0001)
ivw_model$outliers
```


### Cook's distance

In statistics, Cook's distance or Cook's D is a commonly used estimate of the influence of a data point when performing a least-squares regression analysis.[1] In a practical ordinary least squares analysis, Cook's distance can be used in several ways:

1- To indicate influential data points that are particularly worth checking for validity.

2- To indicate regions of the design space where it would be good to be able to obtain more data points. 

**It is named after the American statistician R. Dennis Cook, who introduced the concept in 1977.**

[Refernce](https://en.wikipedia.org/wiki/Cook%27s_distance)

```{r Cook distance1, message=FALSE, warning=FALSE}
par(mfrow = c(2, 2))
model <- lm(data$beta.outcome~data$beta.exposure-1)
plot(model)
par(mfrow = c(1, 1))

cooksD <- cooks.distance(model)
influential <- cooksD[(cooksD > (3 * mean(cooksD, na.rm = TRUE)))]
influential
```




### Run After deleting new outlier: Final Results:

```{r Final Results1, message=FALSE, warning=FALSE}

data <- data[(data$SNP!="rs319748" & data$SNP!="rs624833"),]

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)


#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#scatter plot
p1 <- mr_scatter_plot(res, data); p1[[1]]

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Sensitivity analyses with MendelianRandomization Package

```{r MendelianRandomization1, echo=FALSE, message=FALSE, warning=FALSE}

MRInputObject <- mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome)
IVWObject <- mr_ivw(MRInputObject,
                    model = "default",
                    robust = FALSE,
                    penalized = FALSE,
                    correl = FALSE,
                    weights = "simple",
                    psi = 0,
                    distribution = "normal",
                    alpha = 0.05)
IVWObject

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")
MRAllObject_all
mr_plot(MRAllObject_all)

mr_forest(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
          snp_estimates=FALSE,
          methods = c("ivw", "median", "wmedian", "egger", "maxlik", "mbe"))
###########
data$samplesize.exposure <- " 341427"
data$samplesize.outcome <- "458620"
data$samplesize.exposure <- as.numeric(data$samplesize.exposure)
data$samplesize.outcome <- as.numeric(data$samplesize.outcome)

out <- directionality_test(data)
knitr::kable(out)

mr_steiger(
  p_exp = data$pval.exposure, 
  p_out = data$pval.outcome, 
  n_exp = data$samplesize.exposure, 
  n_out = data$samplesize.outcome, 
  r_xxo = 1, 
  r_yyo = 1,
  r_exp=0,
  r_out = 0.5
)
```

### Working with  MRraps

```{r MRraps1, echo=FALSE, message=FALSE, warning=FALSE}

library(mr.raps)
mr.raps.simple(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome, diagnosis = FALSE)
mr.raps.all(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome)
mr_lasso(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_cML(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), MA = TRUE, DP = FALSE,, n = 458620)
mr_divw(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_mbe(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), weighting="weighted", stderror="delta",iterations=100)
mr_funnel(MRInputObject, CI = TRUE)
mr_plot(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
        error = TRUE, orientate = FALSE, line = "ivw", interactive = FALSE)
```


## Exposure 3


### Introduction

* **Title:** Investigating the causality between **Cigarettes per day on GD**

    + Exposure:Cigarettes per day, [GSCAN consortium](https://genome.psych.umn.edu/index.php/GSCAN) | Reference paper:  [Liu Paper 2019](https://www.nature.com/articles/s41588-018-0307-5) | GWAS ID:[GCST007459](https://www.ebi.ac.uk/gwas/studies/GCST007459) 
    + **Sample size**: 377,334 
    + Outcome: Graves disease (GD) Disease, [NBC Human Database](https://humandbs.biosciencedbc.jp/en/hum0197-v3-220) | Reference paper: [Saori Sakaue Paper 2022](https://pubmed.ncbi.nlm.nih.gov/34594039/) | GWAS ID: [GCST90018847](https://www.ebi.ac.uk/gwas/studies/GCST90018847)
    + **Sample size**: 458,620 , **Number of cases**: 1,678 , **Number of controls**: 456,942
    + Ancestry: **European** 
    

### Data Preparation
1- **Number of total SNPs in exposure:** 5,213 SNPs

2- **Number of SNPs exposure with p-value <** $10^-8$ = 2,129 SNPs

3- **Number of SNPs exposure after clumping ** = 23 SNPs

4- **Number of total SNPs in outcome**: 4,943 SNPs

5- **Number of common variants between exposure and outcome**: 21 SNPs("rs7951365" and "rs2273500"  have been eliminated)

6- **Number of SNPs after replacing proxies**: 1 SNPs from [NIH LDproxy database](https://ldlink.nih.gov/?tab=ldproxy) according to EUR ancestry have been selected: We replace "rs7951365" by rs7933830 with R2 1. So, 22 SNPs remained.   

7- **Number of SNPs after harmonization (action=2) ** = 21 SNPs (Removing the following SNPs for being palindromic with intermediate allele frequencies:rs787362)

8- **Number of SNPs after removing HLA region with exploring in [HLA Genes, Nomenclature](https://hla.alleles.org/genes/index.html)** = 21 SNP 

9- **Number of SNPs after removing those that have MAF < 0.01** = 21 SNPs 

10- Checking pleiotropy by [PhenoScanner](http://www.phenoscanner.medschl.cam.ac.uk/):

How many SNPs have been eliminated after checking the PhenoScanner website: 21 SNPs



### Checking weakness of the instruments:

```{r Checking weekness2, message=FALSE, warning=FALSE}
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_CigarettesPerDay_Graves_AFTER_LD.txt")
data$F<-(data$beta.exposure/data$se.exposure)^2
summary(data$F)
```
How many SNPs have been eliminated with checking the weakness: 0 SNP


### RUN an initial MR:

```{r RUN an initial MR2, message=FALSE, warning=FALSE}
res<-mr(data)
res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)	
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Testing Outlier with PRESSO test

```{r PRESSO2,  message=FALSE, warning=FALSE}

#Radial test
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data <- fread("data_CigarettesPerDay_Graves_AFTER_LD.txt")
data <- as.data.frame(data)

mr_presso(BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", SdOutcome = "se.outcome", SdExposure = "se.exposure",
          OUTLIERtest = TRUE, DISTORTIONtest = TRUE, data = data, NbDistribution = 1000,  SignifThreshold = 0.05)

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```


### Studentized residuals:

```{r Studentized residuals2, message=FALSE, warning=FALSE}
reg_1<-lm(data$beta.outcome~data$beta.exposure-1)
data$st_1<-rstandard(reg_1)

#Histogram plot
hist(data$st_1)
```

### Radial test

```{r Radial2, message=FALSE, warning=FALSE}
#Radial test: 

radial_data<-format_radial(data$beta.exposure, data$beta.outcome,
                           data$se.exposure, data$se.outcome,
                           data$SNP)
ivw_model<-ivw_radial(radial_data, 0.05/nrow(radial_data),3,0.0001)
ivw_model$outliers
```


### Cook's distance

In statistics, Cook's distance or Cook's D is a commonly used estimate of the influence of a data point when performing a least-squares regression analysis.[1] In a practical ordinary least squares analysis, Cook's distance can be used in several ways:

1- To indicate influential data points that are particularly worth checking for validity.

2- To indicate regions of the design space where it would be good to be able to obtain more data points. 

**It is named after the American statistician R. Dennis Cook, who introduced the concept in 1977.**

[Refernce](https://en.wikipedia.org/wiki/Cook%27s_distance)

```{r Cook distance2,, echo=TRUE, message=FALSE, warning=FALSE}
par(mfrow = c(2, 2))
model <- lm(data$beta.outcome~data$beta.exposure-1)
plot(model)
par(mfrow = c(1, 1))

cooksD <- cooks.distance(model)
influential <- cooksD[(cooksD > (3 * mean(cooksD, na.rm = TRUE)))]
influential
```




### Run After deleting new outlier: Final Results:

```{r Final Results2,, message=FALSE, warning=FALSE}


data <- data[(data$SNP!="rs7431710" & data$SNP!="rs75494138" & data$SNP!="rs2386571"& data$SNP!="rs11852372" 
              & data$SNP!="rs2072659" & data$SNP!="rs3025383" & data$SNP!="rs34406232" & data$SNP!="rs11725618"
              & data$SNP!="rs2084533" & data$SNP!="rs7951365"),]

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)


#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#scatter plot
p1 <- mr_scatter_plot(res, data); p1[[1]]

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Sensitivity analyses with MendelianRandomization Package

```{r MendelianRandomization2, message=FALSE, warning=FALSE}

MRInputObject <- mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome)
IVWObject <- mr_ivw(MRInputObject,
                    model = "default",
                    robust = FALSE,
                    penalized = FALSE,
                    correl = FALSE,
                    weights = "simple",
                    psi = 0,
                    distribution = "normal",
                    alpha = 0.05)
IVWObject

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")
MRAllObject_all
mr_plot(MRAllObject_all)

mr_forest(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
          snp_estimates=FALSE,
          methods = c("ivw", "median", "wmedian", "egger", "maxlik", "mbe"))
###########
data$samplesize.exposure <- " 377334"
data$samplesize.outcome <- "458620"
data$samplesize.exposure <- as.numeric(data$samplesize.exposure)
data$samplesize.outcome <- as.numeric(data$samplesize.outcome)

out <- directionality_test(data)
knitr::kable(out)

mr_steiger(
  p_exp = data$pval.exposure, 
  p_out = data$pval.outcome, 
  n_exp = data$samplesize.exposure, 
  n_out = data$samplesize.outcome, 
  r_xxo = 1, 
  r_yyo = 1,
  r_exp=0,
  r_out = 0.5
)
```

### Working with  MRraps

```{r MRraps2, message=FALSE, warning=FALSE}

library(mr.raps)
mr.raps.simple(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome, diagnosis = FALSE)
mr.raps.all(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome)
mr_lasso(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_cML(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), MA = TRUE, DP = FALSE,, n = 458620)
mr_divw(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_mbe(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), weighting="weighted", stderror="delta",iterations=100)
mr_funnel(MRInputObject, CI = TRUE)
mr_plot(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
        error = TRUE, orientate = FALSE, line = "ivw", interactive = FALSE)
```


## Exposure 4

### Introduction

* **Title:** Investigating the causality between **Life time Smoking on GD**

    + Exposure:Life time smoking, [UK biobank consortium](https://www.ukbiobank.ac.uk/) | Reference paper:  [Wootton RE 2019](https://pubmed.ncbi.nlm.nih.gov/31689377/) | GWAS ID:[GCST009096](https://www.ebi.ac.uk/gwas/studies/GCST009096) 
    + **Sample size**: 462,690
    + Outcome: Graves disease (GD) Disease, [NBC Human Database](https://humandbs.biosciencedbc.jp/en/hum0197-v3-220) | Reference paper: [Saori Sakaue Paper 2022](https://pubmed.ncbi.nlm.nih.gov/34594039/) | GWAS ID: [GCST90018847](https://www.ebi.ac.uk/gwas/studies/GCST90018847)
    + **Sample size**: 458,620 , **Number of cases**: 1,678 , **Number of controls**: 456,942
    + Ancestry: **European** 
    

### Data Preparation
1- **Number of total SNPs in exposure:** 52,663 SNPs

2- **Number of SNPs exposure with p-value <** $10^-8$ = 10,413 SNPs

3- **Number of SNPs exposure after clumping ** = 126 SNPs

4- **Number of total SNPs in outcome**: 50,272 SNPs

5- **Number of common variants between exposure and outcome**: 118 SNPs("rs13016665", "rs12623702", "rs73220544" "rs317021",  "rs10823968", "rs12244388", "rs7297175" and  "rs369230"  have been eliminated)

6- **Number of SNPs after replacing proxies**: 3 SNPs from [NIH LDproxy database](https://ldlink.nih.gov/?tab=ldproxy) according to EUR ancestry have been selected: We replace "rs12623702", "rs73220544" and "rs7297175" by rs1477031, rs62280815 and rs11171739 with R2 0.99, 1 and 0.98.So, 121 SNPs remained.   

7- **Number of SNPs after harmonization (action=2) ** = 115 SNPs (Removing the following SNPs for incompatible alleles:
rs73220544 and Removing the following SNPs for being palindromic with intermediate allele frequencies:
rs10922907, rs2401924, rs2678670, rs3769949, rs6692614)

8- **Number of SNPs after removing HLA region with exploring in [HLA Genes, Nomenclature](https://hla.alleles.org/genes/index.html)** = 115 SNP 

9- **Number of SNPs after removing those that have MAF < 0.01** = 115 SNPs 

10- Checking pleiotropy by [PhenoScanner](http://www.phenoscanner.medschl.cam.ac.uk/):

How many SNPs have been eliminated after checking the PhenoScanner website: 115 SNPs



### Checking weakness of the instruments:

```{r Checking weekness3, message=FALSE, warning=FALSE}
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data<-fread("data_lifetime_graves_AFTER_LD.txt")
data$F<-(data$beta.exposure/data$se.exposure)^2
summary(data$F)
```
How many SNPs have been eliminated with checking the weakness: 0 SNP


### RUN an initial MR:

```{r RUN an initial MR3, message=FALSE, warning=FALSE}
res<-mr(data)
res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)	
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Testing Outlier with PRESSO test

```{r PRESSO3, message=FALSE, warning=FALSE}

#Radial test
setwd("E:\\MRW\\Graves_4 Smoking\\smoking-eur\\smoking_graves")
data<-fread("data_lifetime_graves_AFTER_LD.txt")
data <- as.data.frame(data)

mr_presso(BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", SdOutcome = "se.outcome", SdExposure = "se.exposure",
          OUTLIERtest = TRUE, DISTORTIONtest = TRUE, data = data, NbDistribution = 1000,  SignifThreshold = 0.05)

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)

#scatter plot
p1 <- mr_scatter_plot(res, data)
p1[[1]]

#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```


### Studentized residuals:

```{r Studentized residuals3,  message=FALSE, warning=FALSE}
reg_1<-lm(data$beta.outcome~data$beta.exposure-1)
data$st_1<-rstandard(reg_1)

#Histogram plot
hist(data$st_1)
```

### Radial test

```{r Radial3, message=FALSE, warning=FALSE}
#Radial test: 

radial_data<-format_radial(data$beta.exposure, data$beta.outcome,
                           data$se.exposure, data$se.outcome,
                           data$SNP)
ivw_model<-ivw_radial(radial_data, 0.05/nrow(radial_data),3,0.0001)
ivw_model$outliers
```


### Cook's distance

In statistics, Cook's distance or Cook's D is a commonly used estimate of the influence of a data point when performing a least-squares regression analysis.[1] In a practical ordinary least squares analysis, Cook's distance can be used in several ways:

1- To indicate influential data points that are particularly worth checking for validity.

2- To indicate regions of the design space where it would be good to be able to obtain more data points. 

**It is named after the American statistician R. Dennis Cook, who introduced the concept in 1977.**

[Refernce](https://en.wikipedia.org/wiki/Cook%27s_distance)

```{r Cook distance3, message=FALSE, warning=FALSE}
par(mfrow = c(2, 2))
model <- lm(data$beta.outcome~data$beta.exposure-1)
plot(model)
par(mfrow = c(1, 1))

cooksD <- cooks.distance(model)
influential <- cooksD[(cooksD > (3 * mean(cooksD, na.rm = TRUE)))]
influential
```




### Run After deleting new outlier: Final Results:

```{r Final Results3, message=FALSE, warning=FALSE}


data <- data[(data$SNP!="rs4473348"  & data$SNP!="rs57611503" & data$SNP!="rs13009008"  & data$SNP!="rs8042849"
                 & data$SNP!="rs74086911" & data$SNP!="rs329120" & data$SNP!="rs113382419" & data$SNP!="rs6119897" 
                 & data$SNP!="rs112282219" & data$SNP!="rs6741228" & data$SNP!="rs72674867" & data$SNP!="rs28635466"
                 & data$SNP!="rs71367545" & data$SNP!="rs7077678" & data$SNP!="rs34866095" & data$SNP!="rs10282292"
                 & data$SNP!="rs7569203" & data$SNP!="rs3811038" & data$SNP!="rs11768481" & data$SNP!="rs2675638"),]

res<-mr(data)

res
plot(data$beta.exposure,data$beta.outcome)
text(data$beta.exposure,                                
     data$beta.outcome,
     labels = data$SNP,
     pos = 4)


#Heterogeneity testing
mr_heterogeneity<- mr_heterogeneity(data); mr_heterogeneity

#pleiotropy testing
mr_pleiotropy_test<- mr_pleiotropy_test(data); mr_pleiotropy_test

#scatter plot
p1 <- mr_scatter_plot(res, data); p1[[1]]

#plot of single SNP MR:
res_single <- mr_singlesnp(data); p2 <- mr_forest_plot(res_single); p2[[1]]

#plot of LOO:
res_loo <- mr_leaveoneout(data); p3 <- mr_leaveoneout_plot(res_loo); p3[[1]]

#Funnel plot
p4 <- mr_funnel_plot(res_single); p4[[1]]
```



### Sensitivity analyses with MendelianRandomization Package

```{r MendelianRandomization3, message=FALSE, warning=FALSE}

MRInputObject <- mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome)
IVWObject <- mr_ivw(MRInputObject,
                    model = "default",
                    robust = FALSE,
                    penalized = FALSE,
                    correl = FALSE,
                    weights = "simple",
                    psi = 0,
                    distribution = "normal",
                    alpha = 0.05)
IVWObject

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")
MRAllObject_all
mr_plot(MRAllObject_all)

mr_forest(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
          snp_estimates=FALSE,
          methods = c("ivw", "median", "wmedian", "egger", "maxlik", "mbe"))
###########
data$samplesize.exposure <- "52663"
data$samplesize.outcome <- "50272"
data$samplesize.exposure <- as.numeric(data$samplesize.exposure)
data$samplesize.outcome <- as.numeric(data$samplesize.outcome)

out <- directionality_test(data)
knitr::kable(out)

mr_steiger(
  p_exp = data$pval.exposure, 
  p_out = data$pval.outcome, 
  n_exp = data$samplesize.exposure, 
  n_out = data$samplesize.outcome, 
  r_xxo = 1, 
  r_yyo = 1,
  r_exp=0,
  r_out = 0.5
)
```

### Working with  MRraps

```{r MRraps3, message=FALSE, warning=FALSE}

library(mr.raps)
mr.raps.simple(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome, diagnosis = FALSE)
mr.raps.all(data$beta.exposure,data$beta.outcome, data$se.exposure, data$se.outcome)
mr_lasso(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_cML(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), MA = TRUE, DP = FALSE,, n = 458620)
mr_divw(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome))
mr_mbe(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome), weighting="weighted", stderror="delta",iterations=100)
mr_funnel(MRInputObject, CI = TRUE)
mr_plot(mr_input(bx = data$beta.exposure,bxse = data$se.exposure,by = data$beta.outcome,byse = data$se.outcome),
        error = TRUE, orientate = FALSE, line = "ivw", interactive = FALSE)
```


